Living-Snow-Project/LivingSnowProject
# feat: background record uploading

### What Changed
A background upload task was added for records that fail during the first upload attempt.

The background task is created in `useRecordReducer.ts` so that the task function can have access to the reducer `dispatch` function. Without this, the records will be uploaded in the background but will show as pending on the timeline until the user manually refreshes the page.

The task is registered if the first upload attempt fails and unregisters itself when `loadPendingRecords()` returns an empty list. 

### Testing ( @chris-chapin I accidentally uploaded a test entry with photos of a cat to the production server...)
This was tested locally using a containerized imitation of the API I quickly built here: [living-snow-local-testapi](https://github.com/tccw/living-snow-local-testapi). I simply tried to provide the same responses the front-end expected.

I used the expo-go app on Android and the iOS simulator provided by XCode to test this. Real background app refreshes on iPhones are somewhat unpredictable so I utilized the simulator and [these instructions](https://docs.expo.dev/versions/latest/sdk/background-fetch/#triggering-background-fetches) to test on iOS.

To test on Android, add `minimumInterval: 1` to the configuration call in line 215 of `useRecordReducer.ts`. This makes the interval between backgrounding the app and the task running very short.

### Manual Testing
**case**: record fails to upload
**expected outcome**: both record and photos are not uploaded and the background task is registered ✅ 

**case**: record uploads but photos fail
**expected outcome**: photos are cached and a background task is registered to upload the photos later✅ 

**case**: multiple records fail to upload
**expected outcome**: photos are cached and all records are uploaded once back online✅ 

**case**: record fails to upload, and the user has background app refresh/tasks disabled
**expected outcome**: records are cached (as usual) and a new Alert pops up to tell the user that they should enable Background App Refresh ✅ 

Closes #17
